pkg/mls/store/store.go

```diff
@@ -157,8 +171,189 @@ func (s *Store) RevokeInstallation(ctx context.Context, installationId []byte) e
+	message := GroupMessage{
+		Data: data,
+	}
+
+	var id uint64
+	err := s.db.QueryRow("INSERT INTO group_messages (group_id, data, group_id_data_hash) VALUES (?, ?, ?) RETURNING id", groupId, data, sha256.Sum256(append(groupId, data...))).Scan(&id)
+	if err != nil {
+		if strings.Contains(err.Error(), "duplicate key value violates unique constraint") {
+			return nil, NewAlreadyExistsError(err)
+		}
+		return nil, err
+	}
+
+	err = s.db.NewSelect().Model(&message).Where("id = ?", id).Scan(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &message, nil
+}
+
+func (s *Store) InsertWelcomeMessage(ctx context.Context, installationId []byte, data []byte) (*WelcomeMessage, error) {
+	message := WelcomeMessage{
+		Data: data,
+	}
+
+	var id uint64
+	err := s.db.QueryRow("INSERT INTO welcome_messages (installation_key, data, installation_key_data_hash) VALUES (?, ?, ?) RETURNING id", installationId, data, sha256.Sum256(append(installationId, data...))).Scan(&id)
+	if err != nil {
+		if strings.Contains(err.Error(), "duplicate key value violates unique constraint") {
+			return nil, NewAlreadyExistsError(err)
+		}
+		return nil, err
+	}
+
+	err = s.db.NewSelect().Model(&message).Where("id = ?", id).Scan(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &message, nil
+}
+
+func (s *Store) QueryGroupMessagesV1(ctx context.Context, req *mlsv1.QueryGroupMessagesRequest) (*mlsv1.QueryGroupMessagesResponse, error) {
+	msgs := make([]*GroupMessage, 0)
+
+	if len(req.GroupId) == 0 {
+		return nil, errors.New("group is required")
+	}
+
+	q := s.db.NewSelect().
+		Model(&msgs).
+		Where("group_id = ?", req.GroupId)
+
+	direction := mlsv1.SortDirection_SORT_DIRECTION_DESCENDING
+	if req.PagingInfo != nil && req.PagingInfo.Direction != mlsv1.SortDirection_SORT_DIRECTION_UNSPECIFIED {
+		direction = req.PagingInfo.Direction
+	}
+	switch direction {
+	case mlsv1.SortDirection_SORT_DIRECTION_DESCENDING:
+		q = q.Order("id DESC")
+	case mlsv1.SortDirection_SORT_DIRECTION_ASCENDING:
+		q = q.Order("id ASC")
+	}
+
+	pageSize := maxPageSize
+	if req.PagingInfo != nil && req.PagingInfo.Limit > 0 && req.PagingInfo.Limit <= maxPageSize {
+		pageSize = int(req.PagingInfo.Limit)
+	}
+	q = q.Limit(pageSize)
+
+	if req.PagingInfo != nil && req.PagingInfo.IdCursor != 0 {
+		if direction == mlsv1.SortDirection_SORT_DIRECTION_ASCENDING {
+			q = q.Where("id > ?", req.PagingInfo.IdCursor)
+		} else {
+			q = q.Where("id < ?", req.PagingInfo.IdCursor)
+		}
+	}
+
+	err := q.Scan(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	messages := make([]*mlsv1.GroupMessage, 0, len(msgs))
+	for _, msg := range msgs {
+		messages = append(messages, &mlsv1.GroupMessage{
+			Version: &mlsv1.GroupMessage_V1_{
+				V1: &mlsv1.GroupMessage_V1{
+					Id:        msg.Id,
+					CreatedNs: uint64(msg.CreatedAt.UnixNano()),
+					GroupId:   msg.GroupId,
+					Data:      msg.Data,
+				},
+			},
+		})
+	}
+
+	pagingInfo := &mlsv1.PagingInfo{Limit: 0, IdCursor: 0, Direction: direction}
```

Comment:
We should use the original limit, so that the caller can just pass back the `PagingInfo` into the next request and not be surprised by the number of results.